server/src/main/java/io/crate/metadata/doc/DocTableInfo.java

```diff
@@ -492,4 +499,23 @@ public String toString() {
     public Collection<ColumnIdent> notNullColumns() {
         return notNullColumns;
     }
+
+    /**
+     * Starting from 5.5 column OID-s are used as source keys.
+     * Even of 5.5, there are no OIDs (and thus no source key rewrite happening) for:
+     * <ul>
+     *  <li>OBJECT (IGNORED) sub-columns</li>
+     *  <li>Empty arrays, or arrays with only null values</li>
+     * </ul>
+     */
+    public Function<String, String> lookupNameBySourceKey() {
+        if (versionCreated.onOrAfter(Version.V_5_5_0)) {
+            return oidOrName -> {
+                String leafName = leafNamesByOid.get(oidOrName);
+                return leafName != null ? leafName : oidOrName;
+            };
+        } else {
+            return Function.identity();
+        }
+    }
 }
```

Comment:
This passed tests even before we write oids to source since by default References have OID_UNASSIGNED and they fall back to "take as is" behavior.